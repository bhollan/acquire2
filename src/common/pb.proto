syntax = "proto3";

message PB {
  enum ErrorCode {
    NOT_USING_LATEST_VERSION = 0;
    INTERNAL_SERVER_ERROR = 1;
    INVALID_MESSAGE_FORMAT = 2;
    INVALID_USERNAME = 3;
    MISSING_PASSWORD = 4;
    PROVIDED_PASSWORD = 5;
    INCORRECT_PASSWORD = 6;
    INVALID_MESSAGE = 7;
    COULD_NOT_CONNECT = 8;
  }

  enum GameBoardType {
    LUXOR = 0;
    TOWER = 1;
    AMERICAN = 2;
    FESTIVAL = 3;
    WORLDWIDE = 4;
    CONTINENTAL = 5;
    IMPERIAL = 6;
    NOTHING = 7;
    NOTHING_YET = 8;
    CANT_PLAY_EVER = 9;
    I_HAVE_THIS = 10;
    WILL_PUT_LONELY_TILE_DOWN = 11;
    HAVE_NEIGHBORING_TILE_TOO = 12;
    WILL_FORM_NEW_CHAIN = 13;
    WILL_MERGE_CHAINS = 14;
    CANT_PLAY_NOW = 15;
    MAX = 16;
  }

  enum GameMode {
    UNKNOWN = 0; // not used, here because the first enum value must be zero in proto3
    SINGLES_1 = 1;
    SINGLES_2 = 2;
    SINGLES_3 = 3;
    SINGLES_4 = 4;
    SINGLES_5 = 5;
    SINGLES_6 = 6;
    TEAMS_2_VS_2 = 7;
    TEAMS_2_VS_2_VS_2 = 8;
    TEAMS_3_VS_3 = 9;
  }

  enum PlayerArrangementMode {
    VERSION_1 = 0;
    RANDOM_ORDER = 1;
    EXACT_ORDER = 2;
    SPECIFY_TEAMS = 3;
  }

  message GameSetup {
    message Position {
      int32 user_id = 1;
      bool is_host = 2;
      bool approves_of_game_setup = 3;
    }

    int32 game_id = 1;
    int32 game_display_number = 2;
    GameMode game_mode = 4;
    PlayerArrangementMode player_arrangement_mode = 5;
    repeated Position positions = 6;
    repeated int32 game_board = 7;
  }

  message GameState {
    message RevealedTileRackTile {
      int32 tile = 1;
      int32 player_id_belongs_to = 2;
    }

    GameAction game_action = 1;
    int64 timestamp = 2 [ jstype = JS_NUMBER ];
    repeated RevealedTileRackTile revealed_tile_rack_tiles = 3;
    repeated int32 revealed_tile_bag_tiles = 4;
    int32 player_id_with_playable_tile_plus_one = 5; // 1 more than actual player_id
  }

  message GameSetupAction {
    message JoinGame {
    }

    message UnjoinGame {
    }

    message ApproveOfGameSetup {
    }

    message ChangeGameMode {
      GameMode game_mode = 1;
    }

    message ChangePlayerArrangementMode {
      PlayerArrangementMode player_arrangement_mode = 1;
    }

    message SwapPositions {
      int32 position1 = 1;
      int32 position2 = 2;
    }

    message KickUser {
      int32 user_id = 1;
    }

    JoinGame join_game = 1;
    UnjoinGame unjoin_game = 2;
    ApproveOfGameSetup approve_of_game_setup = 3;
    ChangeGameMode change_game_mode = 4;
    ChangePlayerArrangementMode change_player_arrangement_mode = 5;
    SwapPositions swap_positions = 6;
    KickUser kick_user = 7;
  }

  message GameSetupChange {
    message UserAdded {
      int32 user_id = 1;
    }

    message UserRemoved {
      int32 user_id = 1;
    }

    message UserApprovedOfGameSetup {
      int32 user_id = 1;
      bool approved_by_everybody = 2;
      repeated int32 final_user_ids = 3;
    }

    message GameModeChanged {
      GameMode game_mode = 1;
    }

    message PlayerArrangementModeChanged {
      PlayerArrangementMode player_arrangement_mode = 1;
    }

    message PositionsSwapped {
      int32 position1 = 1;
      int32 position2 = 2;
    }

    message UserKicked {
      int32 user_id = 1;
    }

    UserAdded user_added = 1;
    UserRemoved user_removed = 2;
    UserApprovedOfGameSetup user_approved_of_game_setup = 3;
    GameModeChanged game_mode_changed = 4;
    PlayerArrangementModeChanged player_arrangement_mode_changed = 5;
    PositionsSwapped positions_swapped = 6;
    UserKicked user_kicked = 7;
  }

  message GameAction {
    message StartGame {
    }

    message PlayTile {
      int32 tile = 1;
    }

    message SelectNewChain {
      GameBoardType chain = 1;
    }

    message SelectMergerSurvivor {
      GameBoardType chain = 1;
    }

    message SelectChainToDisposeOfNext {
      GameBoardType chain = 1;
    }

    message DisposeOfShares {
      int32 trade_amount = 1;
      int32 sell_amount = 2;
    }

    message PurchaseShares {
      repeated GameBoardType chains = 1;
      bool end_game = 2;
    }

    message GameOver {
    }

    StartGame start_game = 1;
    PlayTile play_tile = 2;
    SelectNewChain select_new_chain = 3;
    SelectMergerSurvivor select_merger_survivor = 4;
    SelectChainToDisposeOfNext select_chain_to_dispose_of_next = 5;
    DisposeOfShares dispose_of_shares = 6;
    PurchaseShares purchase_shares = 7;
    GameOver game_over = 8;
  }

  message MessageToServer {
  }

  message MessageToClient {
  }

  message MessagesToClient {
    repeated MessageToClient messages_to_client = 1;
  }

  message GameReview {
    GameMode game_mode = 1;
    PlayerArrangementMode player_arrangement_mode = 2;
    repeated int32 user_ids = 3;
    repeated string usernames = 4;
    int32 host_user_id = 5;
    repeated int32 tile_bag = 6;
    repeated GameAction game_actions = 7;
    int64 begin_timestamp = 8 [ jstype = JS_NUMBER ];
    repeated int64 game_action_timestamp_offsets = 9 [ jstype = JS_NUMBER ];
    int64 end_timestamp = 10 [ jstype = JS_NUMBER ];
  }
}
