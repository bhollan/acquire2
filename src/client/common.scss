// in svelte's global.css
// BEGIN

.luxor {
  background-color: #ff0000;
  color: #ffffff;
}

.tower {
  background-color: #ffff00;
}

.american {
  background-color: #0000ff;
  color: #ffffff;
}

.festival {
  background-color: #00ff00;
}

.worldwide {
  background-color: #804000;
  color: #ffffff;
}

.continental {
  background-color: #00ffff;
}

.imperial {
  background-color: #ff00ff;
  color: #ffffff;
}

.nothing {
  background-color: #ffffc0;
}

.nothingYet {
  background-color: #000000;
  color: #ffffff;
}

.cantPlayEver {
  background-color: #808080;
  color: #ffffff;
}

.iHaveThis {
  background-color: #ffe0a0;
}

.willPutLonelyTileDown {
  background-color: #c0c0c0;
}

.haveNeighboringTileToo {
  background-color: #c0ffc0;
}

.willFormNewChain {
  background-color: #ffffff;
}

.willMergeChains {
  background: url(./merge.png);
}

.cantPlayNow {
  background-color: #606060;
  color: #ffffff;
}

// END

// in svelte's Username component
.username {
  background-color: #d3d3d3;
}

.hotelButton {
  font-weight: bold;
  padding: 1px;

  &:focus {
    box-shadow: 0px 0px 2px 0px #323232;
  }
}

.invisible {
  visibility: hidden;
}

// in svelte's global.css
// BEGIN

.team1 {
  background-color: #fff7e7;
}

.team2 {
  background-color: #eee7ff;
}

.team3 {
  background-color: #ebffe7;
}

// END
