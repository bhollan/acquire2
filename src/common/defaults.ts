import type { GameState } from './game';
import { PB_GameBoardType } from './pb';

export function defaultGameStateHistory(): GameState[] {
	return [];
}

export function defaultTileRacks(): (number | null)[][] {
	return [];
}

export function defaultTileRack(): (number | null)[] {
	return [null, null, null, null, null, null];
}

export function defaultTileRackTypesList(): (PB_GameBoardType | null)[][] {
	return [];
}

export function defaultTileRackTypes(): (PB_GameBoardType | null)[] {
	return [null, null, null, null, null, null];
}

const defaultGameBoardRow = [
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
];
export function defaultGameBoard(): PB_GameBoardType[][] {
	return [
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
	];
}

export function defaultScoreBoard(): number[][] {
	return [];
}

export function defaultScoreBoardRow(): number[] {
	return [0, 0, 0, 0, 0, 0, 0, 60, 60];
}

export function defaultScoreBoardAvailable(): number[] {
	return [25, 25, 25, 25, 25, 25, 25];
}

export function defaultScoreBoardChainSize(): number[] {
	return [0, 0, 0, 0, 0, 0, 0];
}

export function defaultScoreBoardPrice(): number[] {
	return [0, 0, 0, 0, 0, 0, 0];
}

export function defaultSafeChains(): boolean[] {
	return [false, false, false, false, false, false, false];
}
