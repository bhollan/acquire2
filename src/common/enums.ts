export enum GameActionEnum {
	StartGame,
	PlayTile,
	SelectNewChain,
	SelectMergerSurvivor,
	SelectChainToDisposeOfNext,
	DisposeOfShares,
	PurchaseShares,
	GameOver,
}

export enum GameHistoryMessageEnum {
	TurnBegan,
	DrewPositionTile,
	StartedGame,
	DrewTile,
	HasNoPlayableTile,
	PlayedTile,
	FormedChain,
	MergedChains,
	SelectedMergerSurvivor,
	SelectedChainToDisposeOfNext,
	ReceivedBonus,
	DisposedOfShares,
	CouldNotAffordAnyShares,
	PurchasedShares,
	DrewLastTile,
	ReplacedDeadTile,
	EndedGame,
	NoTilesPlayedForEntireRound,
	AllTilesPlayed,
}

export enum ScoreBoardIndexEnum {
	Luxor,
	Tower,
	American,
	Festival,
	Worldwide,
	Continental,
	Imperial,
	Cash,
	Net,
	Max,
}

export enum TileEnum {
	Unknown = 108,
}
