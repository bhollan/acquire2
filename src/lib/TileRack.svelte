<svelte:options immutable />

<script lang="ts" context="module">
	type TileData = {
		tile: number;
		type: PB_GameBoardType;
		element?: HTMLInputElement;
	};

	const keyboardShortcutToTileIndex = new Map([
		['1', 0],
		['2', 1],
		['3', 2],
		['4', 3],
		['5', 4],
		['6', 5],
	]);
</script>

<script lang="ts">
	import { PB_GameBoardType } from '$common/pb';
	import { gameBoardTypeToCSSClassName, getTileString } from './helpers';

	export let tiles: (number | null)[];
	export let types: (PB_GameBoardType | null)[];
	export let buttonSize: number;
	export let keyboardShortcutsEnabled: boolean;
	export let onTileClicked: (tile: number) => void;

	let allTileData: (TileData | undefined)[] = new Array(6);
	allTileData.fill(undefined);

	$: for (let i = 0; i < tiles.length; i++) {
		const tile = tiles[i];
		const type = types[i];

		if (tile !== null && type !== null) {
			if (allTileData[i]) {
				allTileData[i]!.tile = tile;
				allTileData[i]!.type = type;
			} else {
				allTileData[i] = { tile, type };
			}
		} else {
			allTileData[i] = undefined;
		}
	}

	$: buttonStyle = `width: ${buttonSize}px; height: ${buttonSize}px;`;

	function handleKeydown(event: KeyboardEvent) {
		if (keyboardShortcutsEnabled) {
			const tileIndex = keyboardShortcutToTileIndex.get(event.key);

			if (tileIndex !== undefined) {
				allTileData[tileIndex]?.element?.focus();
			}
		}
	}

	function onClick(tileData: TileData | undefined) {
		if (tileData) {
			onTileClicked(tileData.tile);
		}
	}
</script>

<svelte:window on:keydown={handleKeydown} />

<div class="root" style="font-size: {Math.floor(buttonSize * 0.4)}px">
	{#each allTileData as tileData}
		{#if tileData}
			<input
				type="button"
				class={`hotelButton ${gameBoardTypeToCSSClassName.get(tileData.type)}`}
				style={buttonStyle}
				value={getTileString(tileData.tile)}
				disabled={tileData.type === PB_GameBoardType.CANT_PLAY_EVER ||
					tileData.type === PB_GameBoardType.CANT_PLAY_NOW}
				on:click={() => onClick(tileData)}
				bind:this={tileData.element}
			/>
		{:else}
			<input type="button" class="invisible" style={buttonStyle} value="?" />
		{/if}
	{/each}
</div>

<style>
	.root :not(:last-child) {
		margin-right: 4px;
	}
</style>
