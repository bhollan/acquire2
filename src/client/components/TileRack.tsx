import { List } from 'immutable';
import React from 'react';
import { PB_GameBoardType } from '../../common/pb';
import * as commonStyle from '../common.scss';
import { gameBoardTypeToCSSClassName, getTileString } from '../helpers';
import * as style from './TileRack.scss';

export interface TileRackProps {
  tiles: List<number | null>;
  types: List<PB_GameBoardType | null>;
  buttonSize: number;
  keyboardShortcutsEnabled: boolean;
  onTileClicked: (tile: number) => void;
}

const keyboardShortcutToTileIndex = new Map([
  ['1', 0],
  ['2', 1],
  ['3', 2],
  ['4', 3],
  ['5', 4],
  ['6', 5],
]);

export class TileRack extends React.Component<TileRackProps> {
  buttonRefs: React.RefObject<HTMLInputElement>[];

  constructor(props: TileRackProps) {
    super(props);

    this.buttonRefs = new Array(6);
    for (let i = 0; i < 6; i++) {
      this.buttonRefs[i] = React.createRef();
    }
  }

  shouldComponentUpdate(nextProps: TileRackProps) {
    // everything except keyboardShortcutsEnabled
    return (
      nextProps.tiles !== this.props.tiles ||
      nextProps.types !== this.props.types ||
      nextProps.buttonSize !== this.props.buttonSize ||
      nextProps.onTileClicked !== this.props.onTileClicked
    );
  }

  componentDidMount() {
    window.addEventListener('keydown', this.onWindowKeydown);
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.onWindowKeydown);
  }

  onWindowKeydown = (event: KeyboardEvent) => {
    if (!this.props.keyboardShortcutsEnabled) {
      return;
    }

    const keyName = event.key;

    if (keyboardShortcutToTileIndex.has(keyName)) {
      const tileIndex = keyboardShortcutToTileIndex.get(keyName)!;
      const button = this.buttonRefs[tileIndex].current;
      if (button !== null) {
        button.focus();
      }
    }
  };

  render() {
    const { tiles, types, buttonSize, onTileClicked } = this.props;

    const buttonStyle = {
      width: buttonSize,
      height: buttonSize,
    };

    return (
      <div className={style.root} style={{ fontSize: Math.floor(buttonSize * 0.4) }}>
        {tiles.map((tile, i) => {
          const type = types.get(i, null);

          if (tile !== null && type !== null) {
            const disabled = type === PB_GameBoardType.CANT_PLAY_EVER || type === PB_GameBoardType.CANT_PLAY_NOW;
            return (
              <input
                key={i}
                type={'button'}
                className={`${commonStyle.hotelButton} ${gameBoardTypeToCSSClassName.get(type)}`}
                style={buttonStyle}
                value={getTileString(tile)}
                disabled={disabled}
                ref={this.buttonRefs[i]}
                onClick={() => onTileClicked(tile)}
              />
            );
          } else {
            return <input key={i} type={'button'} className={commonStyle.invisible} style={buttonStyle} value={'?'} ref={this.buttonRefs[i]} />;
          }
        })}
      </div>
    );
  }
}
