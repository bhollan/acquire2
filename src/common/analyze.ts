import fs from 'fs';
import path from 'path';
import { GameHistoryMessageEnum } from './enums';
import { gameFromProtocolBuffer } from './gameSerialization';
import { PB_GameReview } from './pb';
import { getGameHistoryMessageString } from './runGameTestFile';

function main(gameFilePath: string) {
  const gameReviewFileContents = fs.readFileSync(gameFilePath);
  const game = gameFromProtocolBuffer(PB_GameReview.fromBinary(gameReviewFileContents));

  for (let i = 0; i < game.gameStateHistory.length; i++) {
    console.log(i);

    const gameState = game.gameStateHistory[i];
    for (const gameHistoryMessage of gameState.gameHistoryMessages) {
      console.log(getGameHistoryMessageString(gameHistoryMessage));
    }
  }

  console.log();
  console.log();
  console.log();

  for (let i = 0; i < game.gameStateHistory.length; i++) {
    const gameState = game.gameStateHistory[i];
    const mergedChains = gameState.gameHistoryMessages.filter((m) => m.gameHistoryMessage === GameHistoryMessageEnum.ReceivedBonus).length > 0;

    if (mergedChains) {
      console.log(i);

      for (const gameHistoryMessage of gameState.gameHistoryMessages) {
        console.log(getGameHistoryMessageString(gameHistoryMessage));
      }
    }
  }

  updateReviewGamePBBinary(gameReviewFileContents);
}

function updateReviewGamePBBinary(gameReviewFileContents: Buffer) {
  const pathToFile = path.join(__dirname, '../../src/client/reviewGamePBBinary.ts');

  const currentContents = fs.readFileSync(pathToFile).toString();
  const desiredContents = `export const reviewGamePBBinary = new Uint8Array(${JSON.stringify([...gameReviewFileContents])});\n`;

  if (desiredContents !== currentContents) {
    fs.writeFileSync(pathToFile, desiredContents);
  }
}

main(process.argv[2]);
