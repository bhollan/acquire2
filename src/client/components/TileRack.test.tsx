import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { PB_GameBoardType } from '../../common/pb';
import { configureEnzyme } from '../configureEnzyme';
import { TileRack } from './TileRack';

configureEnzyme();

test('renders correctly', () => {
  const onTileClicked = () => {
    // do nothing
  };

  const component = renderer.create(
    <TileRack
      tiles={[1, 28, 55, 82, 92, 40, 71, null, 99, 12, 8, 86, 38, 74]}
      types={[
        PB_GameBoardType.LUXOR,
        PB_GameBoardType.TOWER,
        PB_GameBoardType.AMERICAN,
        PB_GameBoardType.FESTIVAL,
        PB_GameBoardType.WORLDWIDE,
        PB_GameBoardType.CONTINENTAL,
        PB_GameBoardType.IMPERIAL,
        null,
        PB_GameBoardType.WILL_MERGE_CHAINS,
        PB_GameBoardType.WILL_PUT_LONELY_TILE_DOWN,
        PB_GameBoardType.HAVE_NEIGHBORING_TILE_TOO,
        PB_GameBoardType.CANT_PLAY_EVER,
        PB_GameBoardType.WILL_FORM_NEW_CHAIN,
        PB_GameBoardType.CANT_PLAY_NOW,
      ]}
      buttonSize={40}
      keyboardShortcutsEnabled={false}
      onTileClicked={onTileClicked}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('calls onTileClicked with tile', () => {
  const onTileClicked = jest.fn();
  const tileRack = shallow(
    <TileRack tiles={[74]} types={[PB_GameBoardType.LUXOR]} buttonSize={40} keyboardShortcutsEnabled={false} onTileClicked={onTileClicked} />,
  );

  tileRack.find('input').simulate('click');

  expect(onTileClicked.mock.calls.length).toBe(1);
  expect(onTileClicked.mock.calls[0]).toEqual([74]);
});
