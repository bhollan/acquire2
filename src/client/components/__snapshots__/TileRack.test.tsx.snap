// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="root"
  style={
    {
      "fontSize": 16,
    }
  }
>
  <input
    className="hotelButton luxor"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="1B"
  />
  <input
    className="hotelButton tower"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="4B"
  />
  <input
    className="hotelButton american"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="7B"
  />
  <input
    className="hotelButton festival"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="10B"
  />
  <input
    className="hotelButton worldwide"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="11C"
  />
  <input
    className="hotelButton continental"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="5E"
  />
  <input
    className="hotelButton imperial"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="8I"
  />
  <input
    className="invisible"
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="?"
  />
  <input
    className="hotelButton willMergeChains"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="12A"
  />
  <input
    className="hotelButton willPutLonelyTileDown"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="2D"
  />
  <input
    className="hotelButton haveNeighboringTileToo"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="1I"
  />
  <input
    className="hotelButton cantPlayEver"
    disabled={true}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="10F"
  />
  <input
    className="hotelButton willFormNewChain"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="5C"
  />
  <input
    className="hotelButton cantPlayNow"
    disabled={true}
    onClick={[Function]}
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
    type="button"
    value="9C"
  />
</div>
`;
